### ISSUE TITLE
Ward parcellation does not yield connected clusters...

### ISSUE THREAD

**Issue by bthirion** [nilearn/nilearn#2620]
... while it should

Code to reproduce:

```
from nilearn.regions import Parcellations
from nilearn.plotting import view_img
from nilearn.datasets import fetch_neurovault_ids
from nilearn.masking import compute_background_mask

func_img = fetch_neurovault_ids(image_ids=['360527']).images[0]
mask_img = compute_background_mask(func_img)
n_parcels = 200

ward = Parcellations(method='ward', n_parcels=n_parcels, mask=mask_img,
                     standardize=False, smoothing_fwhm=3,
                     memory='nilearn_cache', memory_level=1,
                     verbose=0)
ward.fit(func_img)
view_img(ward.labels_img_, threshold = 0, vmax=2, resampling_interpolation='nearest').open_in_browser()
``` 
--> I think that there is a problem with mask handling.

**Comment by NicolasGensollen** [comment]
@bthirion, not sure this is the issue, but the background mask in your code does not have a single connected component:

```python
from nilearn.image import get_data
from nilearn._utils.ndimage import largest_connected_component
print(  np.sum(get_data(mask_img)) -
        np.sum(largest_connected_component(get_data(mask_img)))
      )
# prints 189
```

Maybe changing the mask computation line with (since the default value for connected is False):
 
```python
mask_img = compute_background_mask(func_img, connected=True)
```
??

**Comment by bthirion** [comment]
Still get the issue.  I'm afraid this may come from sklearn (tbc)...

**Comment by NicolasGensollen** [comment]
@bthirion, maybe I don't understand what you mean by connected clusters. My understanding is that `ward.labels_img_` is basically a 3D array with 201 labels (200 regions and 0 for background). 

We want to check whether each region is connected. So we can, for example, mask `ward.labels_img_` for each unique label value, and then compute the features in the resulting binary array (for example with `scipy.ndimage.label`). If the region for this label value is connected, we should get only two features: background and the region of interest.

If tried the following script, which works when `compute_background_mask` uses `connected=True`, and fails when `connected=False`

Note: Sorry if I misunderstand your point.

```python
from nilearn.regions import Parcellations
from nilearn.plotting import view_img
from nilearn.datasets import fetch_neurovault_ids
from nilearn.masking import compute_background_mask
from nilearn.image import get_data
import numpy as np
from scipy.ndimage import label

func_img = fetch_neurovault_ids(image_ids=['360527']).images[0]
n_parcels = 200

for connected_mask in [True, False]:
    print("--> connected_mask = {0}".format(connected_mask))
    
    mask_img = compute_background_mask(func_img, 
                                       connected=connected_mask)

    ward = Parcellations(method='ward', n_parcels=n_parcels, mask=mask_img,
                         standardize=False, smoothing_fwhm=3,
                         memory='nilearn_cache', memory_level=1,
                         verbose=0)
    ward.fit(func_img)
    lab_img = get_data(ward.labels_img_)
    assert lab_img.shape == mask_img.shape

    unique_labels = set(np.unique(lab_img))
    unique_labels.remove(0) # Remove background label
    assert len(unique_labels) == n_parcels

    tot_voxel_count = 0
    # For each label, compute the binary mask, and label 
    # the features in the binary array. If the region is
    # connected, we should find two regions, one for background
    # and one for the cluster with this label.
    for l in unique_labels:
        label_mask = (lab_img == l) # binary mask for this label
        n_voxel_with_this_label = np.sum(label_mask)
        assert n_voxel_with_this_label > 0
        tot_voxel_count += n_voxel_with_this_label
        
        structure = np.ones((3,3,3), dtype=np.int) # Enforce 26-connectivity
        regions, this_n_labels = label(label_mask.astype(np.int), 
                                       structure=structure)
        if this_n_labels != 1:
            print("Error: {0} connected components for label {1}".format(
                   this_n_labels, l))
    assert tot_voxel_count == np.sum(lab_img != 0)
```

Output:

```
Reading local neurovault data.
No image found on local disk.
getting new batch: http://neurovault.org/api/images/360527
Downloading file: https://neurovault.org/media/images/6618/left-right_button_press.nii.gz
Download succeeded, downloaded to: /home/nicolas/nilearn_data/neurovault/collection_6618/image_360527.nii.gz
Already fetched 1 image
--> connected_mask = True
/home/nicolas/GitRepos/nilearn-fork/nilearn/_utils/cache_mixin.py:295: UserWarning: memory_level is currently set to 0 but a Memory object has been provided. Setting memory_level to 1.
  warnings.warn("memory_level is currently set to 0 but "
--> connected_mask = False
Error: 21 connected components for label 1
Error: 245 connected components for label 2
Error: 11 connected components for label 3
Error: 75 connected components for label 4
Error: 22 connected components for label 5
(...)
Error: 10 connected components for label 187
Error: 3 connected components for label 189
Error: 20 connected components for label 196
Error: 2 connected components for label 197
```

**Comment by bthirion** [comment]
Sorry, I could get back to it yesterday.
You're right, having only one connected component solves the issue --- sorry for my stupid answer !
It's actually a bug in an sklearn routine... I'll post it there soon.

**Comment by Remi-Gau** [comment]
The script from this message is still valid.

https://github.com/nilearn/nilearn/issues/2620#issuecomment-736504457

Not sure what to do about this one: @bthirion

**Comment by bthirion** [comment]
I feel it's one for me. Will try and look at it in the coming days.

**Comment by bthirion** [comment]
interim observation: the neighbor graph seems to be OK. So the issue likely comes from Ward implementation in sklearn.

**Comment by bthirion** [comment]
Let me just provide an updated version
```
from nilearn.regions import Parcellations
from nilearn.plotting import view_img
from nilearn.datasets import fetch_neurovault_ids
from nilearn.masking import compute_background_mask
from nilearn.image import get_data
import numpy as np
from scipy.ndimage import label

func_img = fetch_neurovault_ids(image_ids=['360527']).images[0]
n_parcels = 200

for connected_mask in [False]:
    print("--> connected_mask = {0}".format(connected_mask))
    
    mask_img = compute_background_mask(func_img, 
                                       connected=connected_mask)

    ward = Parcellations(method='ward', n_parcels=n_parcels, mask=mask_img,
                         standardize=False, smoothing_fwhm=3,
                         memory='nilearn_cache', memory_level=1,
                         verbose=0)
    ward.fit(func_img)
    lab_img = get_data(ward.labels_img_)
    assert lab_img.shape == mask_img.shape

    unique_labels = set(np.unique(lab_img))
    unique_labels.remove(0) # Remove background label
    assert len(unique_labels) == n_parcels

    tot_voxel_count = 0
    # For each label, compute the binary mask, and label 
    # the features in the binary array. If the region is
    # connected, we should find two regions, one for background
    # and one for the cluster with this label.
    for l in unique_labels:
        label_mask = (lab_img == l) # binary mask for this label
        n_voxel_with_this_label = np.sum(label_mask)
        assert n_voxel_with_this_label > 0
        tot_voxel_count += n_voxel_with_this_label
        
        structure = np.zeros((3,3,3), dtype=int) # Enforce 6-connectivity 
        structure[1, 1, 1] = structure[0, 1, 1] = structure[2, 1, 1] =structure[1, 0, 1] =structure[1, 2, 1] =structure[1, 1, 0] =structure[1, 1, 2] =1
        regions, this_n_labels = label(label_mask.astype(int), 
                                       structure=structure)
        if this_n_labels != 1:
            print("Error: {0} connected components for label {1}".format(
                   this_n_labels, l))
    assert tot_voxel_count == np.sum(lab_img != 0)
```

**Comment by bthirion** [comment]
The point is that when the graph is not connected at the beginning, sklearn will create some links between the nodes to run the  Ward analysis. I think that this is the only reason why we don't obtain connected clusters.
